src/libpmempool/replica.c

```diff
@@ -561,30 +831,393 @@ check_badblocks(struct pool_set *set, struct poolset_health_status *set_hs)
 
 		for (unsigned p = 0; p < rep->nparts; ++p) {
 			const char *path = PART(rep, p)->path;
+			struct part_health_status *part_hs = &rep_hs->part[p];
 
 			int exists = os_access(path, F_OK) == 0;
 			if (!exists)
 				continue;
 
-			int ret = os_badblocks_check_file(path);
+			if (!part_hs->exist_recovery_file) {
+				/*
+				 * Bad blocks recovery file for this part
+				 * does not exist, so the process of saving
+				 * recovery files was interrupted.
+				 */
+				LOG(1,
+					"bad blocks recovery file for this part does not exist -- '%s'",
+					path);
+				return 1;
+			}
+
+			LOG(1,
+				"reading bad blocks from the recovery file -- '%s'",
+				part_hs->recovery_file);
+
+			ret = replica_part_badblocks_recovery_file_read(path,
+								part_hs);
 			if (ret < 0) {
+				LOG(1,
+					"reading bad blocks from the recovery file failed -- '%s'",
+					part_hs->recovery_file);
+				return -1;
+			}
+
+			if (ret > 0) {
+				LOG(1,
+					"incomplete bad blocks recovery file detected -- '%s'",
+					part_hs->recovery_file);
+				return 1;
+			}
+
+			if (part_hs->bbs.bb_cnt) {
+				LOG(3, "part %u contains %u bad blocks -- '%s'",
+					p, part_hs->bbs.bb_cnt, path);
+			}
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * replica_badblocks_recovery_files_remove -- (internal)
+ *                                       remove all bad blocks recovery files
+ */
+static int
+replica_badblocks_recovery_files_remove(struct pool_set *set,
+					struct poolset_health_status *set_hs)
+{
+	LOG(3, "set %p, set_hs %p", set, set_hs);
+
+	int ret = 0;
+
+	for (unsigned r = 0; r < set->nreplicas; ++r) {
+		struct pool_replica *rep = set->replica[r];
+		struct replica_health_status *rep_hs = set_hs->replica[r];
+
+		/* XXX: not supported yet */
+		if (rep->remote)
+			continue;
+
+		for (unsigned p = 0; p < rep->nparts; ++p) {
+			struct part_health_status *part_hs = &rep_hs->part[p];
+
+			if (!part_hs->exist_recovery_file) {
```

Comment:
Can this ever be true? In the `replica_badblocks_recovery_files_read` you expect all parts to have a recovery file.